### ISSUE TITLE
Number of expired messages

### ISSUE THREAD

**Issue by akabos** [nats-io/nats-streaming-server#1114]
Is there any way to find out how many messages have exceeded the storage `max_age` setting?

**Comment by kozlovic** [comment]
No, there is currently way to know how many messages have expired.

**Comment by tarasmatsyk** [comment]
@kozlovic, sorry for reopening the issue, 
I am wondering if you would be open to a PR that adds totalRemoved messages? 

I see you use `totalCount` and `totalBytes` as main metrics on the storage which is good enough for general monitoring. 
If we could have a cumulative metric of total removed messages it would help us to:
- see how and when clean up kicks in (and how it affects # of total messages in the service)
- get an idea of performance patters so we can tweak storage configuration before we encounter any kind of loses

as an downside, it involves adding one more counter which might seem to edge case specific.

I'd be happy to follow up with a PR if that something that fits into your product vision

**Comment by kozlovic** [comment]
@tarasmatsyk Here is the problem I have with that. NATS Streaming is not a work queue, it is message log based. That is, messages are stored regardless of subscription interest and messages are removed ONLY due to limits, not because they have been consumed (https://docs.nats.io/nats-streaming-concepts/channels/message-log).

This means that messages being removed is not a sign of "message loss", but the only way for messages to be removed. You can't realistically have no limits on a given channel otherwise the storage would grow unbounded.

We could report the number of old messages being dropped to make room for new ones, or messages expiring, but again, I don't think that this is representative of any issue. Also, we would not be persisting this information, so the number would have a meaning only at runtime and would reset on every server restart. I would also probably require a change to the public Store interface so that this new counter is accessible to the server, which is a big change.

With above said, do you still think that there would be value to do that?

**Comment by kozlovic** [comment]
PS: If you need work-queue (and interest based messaging where messages are being removed from the system when they have been ack'ed), you may want to look at JetStream instead (in preview at the moment and included in NATS core in the main branch, but you can find some doc in the JetStream utility repo: https://github.com/nats-io/jetstream).

**Comment by tarasmatsyk** [comment]
@kozlovic, that's a very good comment, appreciate your detailed reply. 

JetStream seems to fit my use case better. 

In regard to Nats streaming, I think the answer would depend on use case. Your answer makes perfect sense, here is a scenario to think about. Once in a while, we get a load increase and suddenly thousands of messages dissapear because of:
- low storage limit
- system not being able to process all logs in the configured message life time

If I accept the fact that such losses are by design, than all is good, however there is no way to debug that losses, messages are just gone and I cannot find a way to figure out when and why except doing time guessing in my mind based on configuration I have. Don't get me wrong, the system is outstanding, but I still think some kind of signal on "clean up" would be helpfull.

To sum up, 
1) I am going to give JetStream a try, it looks promising
2) I think there is some value in "deletion" indicator unless streaming is expected to be "losses tolerant". Honestly, I did not expect such issues to appear and it was kind of funny that autocleanup happened sooner than expected (this is kind of leg shooting because of lack of understanding). I am wondering if anyone else has similar feedback

**Comment by kozlovic** [comment]
> Once in a while, we get a load increase and suddenly thousands of messages disappear because..

A spike in the number of removed messages (say due to expiration) would simply mean that a higher number of messages were produced in a short time and so they would be expiring at "the same time" (and would be seen as a spike in the number of removed messages if that number was plotted regularly). But again that would not give you any indication that there was a problem consuming these messages. Even if the consumers all got the messages without any delay/issue, messages *will* expire and produce the spike in number of messages removed. If anything, it would indicate that a higher than usual number of messages were produced in a small amount of time (causing lots to expire at the same time).

Same applies for the other metric (message count/size). Even if all messages are consumed, messages will be removed based on new messages being added to the log and forcing old messages to be removed. So again, the number of removed messages, and any spike in that number, would have no correlation with the way the applications consume the messages but only how messages are produced.

It's not that I think the system is perfect, far from it, but I fail to see how this metric would help you detect an issue consuming messages. It will only show you that messages were produced faster than usual.

JetStream is expected to be GA by end of this year.

**Comment by akabos** [comment]
@kozlovic 

So you are saying that while it is possible to count the number of messages removed from storage, this number alone does not provide a useful metric. For the metric to be useful for the use case @tarasmatsyk described, it should be the number of removed from storage messages which have never been ACKed.  

Please correct me if I am wrong.

**Comment by kozlovic** [comment]
> it should be the number of removed from storage messages which have never been ACKed.

Again, not really. In NATS Streaming, you can have many different subscriptions for the same channel. So a message may have been consumed and ack'ed by application A but application B that normally consumes these messages may have an issue and is not doing that. If the store was keeping track for a given channel which messages are removed that have not been ack'ed (as in never been sent to a subscription or not ack'ed) then it would not help detect this issue.

It is simply not applicable to NATS Streaming. Adding this new metric would actually require significant changes to both store implementations (and change public Store interface) for little if any value, so you understand that I am not inclined to add that to the product.

**Comment by akabos** [comment]
@kozlovic I see your point and it makes perfect sense. Thanks for taking time to explain.

**Comment by kozlovic** [comment]
@akabos @tarasmatsyk If you want to determine that there is an issue with messages that have been removed while some consumers did not had a chance to consume them, I think you can still do it - with a bit of work. Here is a situation where I run the server with `-mm 5` (max messages 5). I start 2 durable consumers and a durable queue group with 2 members.

After 2 messages I stop one of the durable. Of of the queue member "queuemember1" has a MaxInflight of 1 and does not ack the message that is received, so all further messages go to the second member.

I then publish a total of 10 messages, which means that the 5 oldest messages have been removed. Here is a capture of `streaming/channelsz?channel=foo&subs=1`:
```
{
  "name": "foo",
  "msgs": 5,
  "bytes": 116,
  "first_seq": 6,
  "last_seq": 10,
  "subscriptions": [
    {
      "client_id": "app2",
      "inbox": "_INBOX.oVMivvmXnl1mAenCHLrdo3",
      "ack_inbox": "_INBOX.DeM0X14QL4jXvMj7vvZUNU",
      "durable_name": "dur2",
      "is_durable": true,
      "is_offline": false,
      "max_inflight": 1024,
      "ack_wait": 30,
      "last_sent": 10,
      "pending_count": 0,
      "is_stalled": false
    },
    {
      "client_id": "app1",
      "inbox": "_INBOX.HzfbF6OeRyijdn3O9Q2WhE",
      "ack_inbox": "_INBOX.DeM0X14QL4jXvMj7vvZUMN",
      "durable_name": "dur1",
      "is_durable": true,
      "is_offline": true,
      "max_inflight": 1024,
      "ack_wait": 30,
      "last_sent": 2,
      "pending_count": 0,
      "is_stalled": false
    },
    {
      "client_id": "queuemember1",
      "inbox": "_INBOX.DmXMWm7gtuH8tcZhiX7rxA",
      "ack_inbox": "_INBOX.DeM0X14QL4jXvMj7vvZUOb",
      "queue_name": "dur:queue",
      "is_durable": true,
      "is_offline": false,
      "max_inflight": 1,
      "ack_wait": 3600,
      "last_sent": 1,
      "pending_count": 1,
      "is_stalled": true
    },
    {
      "client_id": "queuemember2",
      "inbox": "_INBOX.RVVW2JniKmbXWnbllFB1Bv",
      "ack_inbox": "_INBOX.DeM0X14QL4jXvMj7vvZUPi",
      "queue_name": "dur:queue",
      "is_durable": true,
      "is_offline": false,
      "max_inflight": 1024,
      "ack_wait": 30,
      "last_sent": 10,
      "pending_count": 0,
      "is_stalled": false
    }
  ]
}
```
Notice at the top how the channel "first_seq" is 6 and "last_seq" is 10, with total "msgs" is 5. Now notice the durable consumer with clientID "app1" and durable name "dur1" shows that it is offile "is_offline: true", and its "last_sent" is 2.

So for a non queue subscriber (where "queue_name" is not present), if "last_sent" is < than the channel "first_seq", it means that this consumer will have "lost" messages, that is, some messages have been removed that have never been sent to this consumer. The amount is "first_seq - last_sent - 1". That is, if the durable consumer was started now, it would receive message 6 to 10, so it really missed message 3, 4, 5.
If you don't care about specific consumer, you would need to get the list of all non queue subscribers, take the consumers' MIN "last_sent" and use that to determine if there have been "lost" messages.

For queue subscribers, I show here on purpose that you can't use the same logic since a member's "last_sent" may be lower than the channel's "first_seq" and yet the group has not missed any message. In the above example, the queue group did not miss any message. So you need to collect queue durables with the same "queue_name" and check the MAX "last_sent" value to check if that value is < than channel's "first_seq" to determine if the group has "lost" messages.

I hope this helps!

**Comment by akabos** [comment]
@kozlovic Oh, that helps a lot. Thank you very much!
